comp/core/autodiscovery/listeners/snmp_deduper.go

```diff
@@ -0,0 +1,188 @@
+		devices:         make([]deviceInfo, 0),
+		pendingServices: make([]pendingService, 0),
+	}
+	deduper.initIPCounter(config)
+	return deduper
+}
+
+func (d *deviceDeduperImpl) initIPCounter(config snmp.ListenerConfig) {
+	for _, config := range config.Configs {
+		ipAddr, ipNet, err := net.ParseCIDR(config.Network)
+		startingIP := ipAddr.Mask(ipNet.Mask)
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+
+		for currentIP := startingIP; ipNet.Contains(currentIP); incrementIP(currentIP) {
+			if ignored := config.IsIPIgnored(currentIP); ignored {
+				continue
+			}
+			count, _ := d.ipsCounter.LoadOrStore(currentIP.String(), 0)
+			d.ipsCounter.Store(currentIP.String(), count.(int)+len(config.Authentications))
+		}
+	}
+}
+
+func (d *deviceDeduperImpl) checkPreviousIPs(deviceIP string) bool {
+	previousIPsDiscovered := true
+	d.ipsCounter.Range(func(key, value any) bool {
+		ip := key.(string)
+		count := value.(int)
+		if count > 0 && minimumIP(ip, deviceIP) == ip {
+			previousIPsDiscovered = false
+			return false
+		}
+		return true
+	})
+
+	return previousIPsDiscovered
+}
+
+func (d *deviceDeduperImpl) addDevice(device deviceInfo) {
+	d.Lock()
+	defer d.Unlock()
+	d.devices = append(d.devices, device)
+}
+
+func (d *deviceDeduperImpl) contains(device deviceInfo) bool {
+	d.Lock()
+	defer d.Unlock()
+	for _, existingDevice := range d.devices {
+		if existingDevice.equal(device) {
+			return true
+		}
+	}
+	return false
+}
+
+func (d *deviceDeduperImpl) addPendingService(svc pendingService) {
+	d.Lock()
+	defer d.Unlock()
+
+	newPendingServices := make([]pendingService, 0)
+
+	for _, pendingSvc := range d.pendingServices {
+		if pendingSvc.device.equal(svc.device) {
+			if minimumIP(pendingSvc.device.IP, svc.device.IP) != svc.device.IP {
+				// current svc IP is greater than the pending one, so we don't add it
+				return
+			}
+			// remove the pending service because it's IP is greater than the new one
+		} else {
+			newPendingServices = append(newPendingServices, pendingSvc)
+		}
+	}
+
+	newPendingServices = append(newPendingServices, svc)
+
+	d.pendingServices = newPendingServices
+}
+
+func (d *deviceDeduperImpl) flushPendingServices() []pendingService {
+	d.Lock()
+	defer d.Unlock()
+	svcToActivate := make([]pendingService, 0)
+	newPendingServices := make([]pendingService, 0)
+
+	for _, pendingSvc := range d.pendingServices {
+		log.Debugf("Checking pending service for device %s", pendingSvc.svc.deviceIP)
+		previousIPsScanned := d.checkPreviousIPs(pendingSvc.svc.deviceIP)
+
+		if previousIPsScanned {
+			log.Debugf("All previous IPs scanned for device %s, activating service", pendingSvc.svc.deviceIP)
+			svcToActivate = append(svcToActivate, pendingSvc)
+		} else {
+			newPendingServices = append(newPendingServices, pendingSvc)
+		}
+	}
+	d.pendingServices = newPendingServices
+	return svcToActivate
+}
+
+func (d *deviceDeduperImpl) removeIP(ip string) bool {
+	d.Lock()
+	defer d.Unlock()
+	count, _ := d.ipsCounter.Load(ip)
```

Comment:
Can we use a map of string to atomic.UInt32 here ? This could help remove the lock here.
We can use the Uint32.Add function https://pkg.go.dev/sync/atomic#Uint32.Add which returns a copy of the new value, and use that new value for comparing to zero

Reply:
Another option would be to continue storing int, but using atomic.AddUint32 which acts directly on integer fields https://pkg.go.dev/sync/atomic#AddUint32